src/Executables/CombineH5/CombineH5.cpp

```diff
@@ -0,0 +1,144 @@
+// main module we just have it be empty
+extern "C" void CkRegisterMainModule(void) {}
+
+// Compares source archives of each file against the 0th volume file. Returns
+// true if each source archive matches the 0th one
+bool check_src_files(const std::vector<std::string>& file_names) {
+  h5::H5File<h5::AccessType::ReadWrite> initial_file(file_names[0], true);
+  const auto& src_archive_object_initial =
+      initial_file.get<h5::SourceArchive>("/src");
+  const std::vector<char>& src_tar_initial =
+      src_archive_object_initial.get_archive();
+  for (size_t i = 1; i < file_names.size(); ++i) {
+    h5::H5File<h5::AccessType::ReadWrite> comparison_file(file_names[i], true);
+    const auto& src_archive_object_compare =
+        comparison_file.get<h5::SourceArchive>("/src");
+    const std::vector<char>& src_tar_compare =
+        src_archive_object_compare.get_archive();
+
+    if (src_tar_initial != src_tar_compare) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void combine_h5(const std::string& file_prefix, const std::string& subfile_name,
+                const std::string& output) {
+  // Parses for and stores all input files to be looped over
+  const std::vector<std::string>& file_names =
+      file_system::glob(file_prefix + "*.h5");
+
+  // Checks that volume data was generated with identical versions of SpECTRE
+  if (!check_src_files(file_names)) {
+    ERROR(
+        "One or more of your files were found to have differing src.tar.gz "
+        "files, meaning that they may be from differing versions of SpECTRE.");
+  }
+
+  // Instantiates the output file and the .vol subfile to be filled with the
+  // combined data later
+  h5::H5File<h5::AccessType::ReadWrite> new_file(output + "0.h5", true);
+  new_file.insert<h5::VolumeData>("/" + subfile_name + ".vol");
+  new_file.close_current_object();
+
+  // std::map to store the sorted volume data
+  std::map<int, std::vector<
+                    std::tuple<size_t, double, std::vector<ElementVolumeData>>>>
+      file_map;
+
+  // Opens the original files, sorts the volume data, and stores it in the
+  // std::map
+  for (size_t i = 0; i < file_names.size(); ++i) {
+    h5::H5File<h5::AccessType::ReadWrite> original_file(
+        file_prefix + std::to_string(i) + ".h5", true);
+
+    const auto& original_volume_file =
+        original_file.get<h5::VolumeData>("/" + subfile_name);
+
+    const auto& sorted_volume_data = original_volume_file.get_data_by_element(
+        std::nullopt, std::nullopt, std::nullopt);
+
+    file_map.insert(
+        std::pair<int, std::vector<std::tuple<size_t, double,
+                                              std::vector<ElementVolumeData>>>>(
+            i, sorted_volume_data));
+    original_file.close_current_object();
+  }
+
+  // Append vectors of ElementVolumeData that were split between H5 files for
+  // a given observation id and value
+  for (size_t i = 1; i < file_names.size(); ++i) {
```

Comment:
This algorithm is also very memory intensive. I think you want to append immediately on read, and you want to know the size of the vector of elements before you start inserting (the reason is that if a `std::vector` needs to expand, it creates a new vector internally, then moves all the data into that from the old buffer). That is, you want to do one sweep where you compute the number of elements (we might want to actually add a function to `VolumeData` to just return you that), then you want to create a `std::vector<ElementVolumeData>` that has enough for all elements (I'm assuming you're now looping over one observation id/time at a time), finally you want to read and immediately emplace_back (with move semantics!) into the new vector. Currently you're storing all volume data at least twice in memory simultaneously, and if the total volume data size is more than ~50GB that simply won't work except for on the biggest supercomputers. Working with such large datasets really requires detailed knowledge of how the underlying data structures store things, which is not trivially or something you're expected to know, but is the sort of thing that separates "coders" from "software engineers" and so I think a good thing to learn! :)

Reply:
Yeah, we created the code in modules first (hence the split between read/write) and I was working on a version that appended immediately on read before making the PR to make it more efficient. However, we were more concerned about reducing runtime when we brought it up at one of the meetings and Saul mentioned runtime would be dictated more by disk read/write speeds and did not fully realise the memory implications at the time, so we didn't finish it. Will finish that version up and replace it with that.

> That is, you want to do one sweep where you compute the number of elements (we might want to actually add a function to VolumeData to just return you that)

Is it necessary to add that function? Since all the other data (i.e. bases, quadratures, extents, etc.) are stored by element, could loop over files and `observation_ids` and add `get_extents.size()` up ? Wouldn't take much memory since the `extents` vector would just be a local variable (with scope restricted to the loop) that only ever stores the vector of one file at a time. Unless, we want to add the function for the future in general and not just post-processing work?

Reply:
I think this was asked on Slack, for now just doing `get_extents().size()` is fine. We can figure out how/if we should elide this in the future. We could write the number of elements as an attribute into the header, for example.